Fermions improvements

jwnys opened the pull request:
Improved version of the FermionOperator2nd.
Improvements:
- diagonal terms are handled as such and no longer generate separate connected elements
- better handling of dtypes (bug fix)
- more tests

gcarleo commented on netket/experimental/operator/_fermions_2nd.py:
```diff
@@ -309,32 +336,37 @@ def get_conn_flattened(self, x, sections, pad=False):
 
     @staticmethod
-    @jit(nopython=True)
+    # @jit(nopython=True)
```
why you cannot jit this anymore?

gcarleo commented:
do you also sum up terms with the same connected elements or is it some further optimization we could do in the future?

jwnys commented:
Well, this PR would sum diagonal terms. For off-diagonal elements it seems hard to determine in advance (without sample dependence) which have the same connected elements. Do you have some examples of what you have in mind? The other PR #1277 would e.g. detect "1^ 2" and "2^ 1" and know that they are conjugates. Since for each sample only one of these terms gives non-zero mels, you can reduce half of the hopping terms, for example.

gcarleo commented:
yes I understand what this PR does so I was asking if had plans to generalize it for off-diagonal terms too :) 
In general there will be several off-diagonal connected elements that land on same x' and that we could sum up to remove x', but yes, this will depend on the configuration x you are computing the matrix elements of... it's hard to it beforehand, you need to do it dynamically. The interesting thing is that there are certainly many cases when the resulting sum of mels landing to the same x' gives 0 (for symmetry reasons) so it is really wasteful sum mel(x,x')*psi(x')/psi(x) over the same x' to then just find out that the sum of the mels is zero!

PhilipVinc commented:
Please rebase on current master in order to fix all CI errors.

gcarleo commented:
tests are failing?

jwnys commented:
> tests are failing?

I need to restructure the code a bit according, as Filippo is doing for LocalOperator. Will be for next week probably.

PhilipVinc commented on netket/experimental/operator/_fermion_operator_2nd.py:
```diff
@@ -248,10 +248,15 @@ def copy(self, *, dtype: Optional[DType] = None):
+        op = FermionOperator2nd(
+            self.hilbert, terms, weights, constant=self._constant, dtype=self.dtype
+        )
+        return op
```
Wether this is inplace/outofplace I believe should depend on what is the use case you envisage:
 - if it should be run automatically internally by the `_setup` method, then it should be inplace as it's just an internal method that you use to setup the operator. In that case it should be marked as internal.
 - If it is very expensive, and you expect that most people don't need it, you can make it part of the public API and then it should be out of place.

My vague feeling here is that we should go for option (1), but I don't have an understanding of how expensive that is, so..

jwnys replied:
My goal was still to remove this again, since it might blow up. Cleaning up unnecessary zeros is already sufficient to keep things light for now..

gcarleo commented:
conflicts?

gcarleo commented on test/operator/test_operator.py:
```diff
@@ -545,70 +545,3 @@ def test_operator_on_subspace():
 
     h4 = nk.operator.Heisenberg(hi, g, acting_on_subspace=0)
     assert h4.acting_on == h1.acting_on
-
```
we don't do openfermion any more?

gcarleo replied:
ah right it was moved sorry

PhilipVinc commented:
I've solved the merge conflicts and rebased on top of the latest changes, and fixed a few things...

The main thing that makes me skeptical is that this implementation is essentially equivalent (well, actually a bit a simplification) of the Localoperator's implementation. 
Eventually we should merge them...

But in the meantime I'm good for this. Let me finish reviewing.

jwnys commented:
> I've solved the merge conflicts and rebased on top of the latest changes, and fixed a few things...
> 
> The main thing that makes me skeptical is that this implementation is essentially equivalent (well, actually a bit a simplification) of the Localoperator's implementation. Eventually we should merge them...
> 
> But in the meantime I'm good for this. Let me finish reviewing.

It looks similar to LocalOperator from the fermionic operator side, but the implementation uses non-local Jordan-Wigner Z strings (in the get_conn computation). To my understanding, this would hinder merging them.

PhilipVinc commented:
No! you're right! I always forget the JW mapping hidden behind those things..

jwnys commented:
I should add that an efficient implementation of the PauliStrings and its ladder operators (#1206) will impact this implementation significantly.

PhilipVinc commented:
what do you mean?

jwnys commented:
Since we do a Jordan-Wigner transformation, this implementation maps the fermion operators onto PauliStrings. However, the creation and annihilation operators require X \pm iY. Once PauliStrings has improved, I'd like to rewrite the backend of FermionOperator2nd to work with that under the hood, rather than having this custom get_conn. From the API point of view, nothing would change.
